src/Core/Domain/Product/ValueObject/ProductType.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Domain\Product\ValueObject;
+
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductConstraintException;
+
+final class ProductType
+{
+    /**
+     * Standard product
+     */
+    const TYPE_STANDARD = 0;
+
+    /**
+     * A pack consists multiple units of product.
+     */
+    const TYPE_PACK = 1;
+
+    /**
+     * Items that are not in physical form and can be sold without requiring any shipping
+     * E.g. downloadable photos, videos, software, services etc.
+     */
+    const TYPE_VIRTUAL = 2;
```

Comment:
What about "product with combinations" ? Isn't it a type too ?

Reply:
https://github.com/PrestaShop/PrestaShop/blob/develop/classes/Product.php#L549

I think only these values are handled in legacy

Reply:
Inconsistency of legacy code.

If it's not too complicated, I suggest we handle correctly it in our modern code (so I suggest creating TYPE_WITH_COMBINATIONS) and let the handler do the translation between modern code and legacy code.

Advantage: easier and cleaner if you work in the controller layer, closer to domain and user knowledge
Disadvantage: having modern and legacy code use different domain model / structure might create confusion and bugs